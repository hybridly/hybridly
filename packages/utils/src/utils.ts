import baseMerge from 'deepmerge'
export { default as clone } from 'lodash.clonedeep'

export function random(length: number = 10): string {
	const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'

	let str = ''
	for (let i = 0; i < length; i++) {
		str += chars.charAt(Math.floor(Math.random() * chars.length))
	}

	return str
}

/** Simple pattern matching util. */
export function match<TValue extends string | number = string, TReturnValue = unknown>(
	value: TValue,
	lookup: Record<TValue | 'default', TReturnValue | ((...args: any[]) => TReturnValue)>,
	...args: any[]
): TReturnValue | Promise<TReturnValue> {
	if (value in lookup || 'default' in lookup) {
		const returnValue = value in lookup
			? lookup[value]
			: lookup.default

		return typeof returnValue === 'function' ? returnValue(...args) : returnValue
	}

	const handlers = Object.keys(lookup)
		.map((key) => `"${key}"`)
		.join(', ')

	const error = new Error(`Tried to handle "${value}" but there is no handler defined. Only defined handlers are: ${handlers}.`)

	throw error
}

export function debounce<F extends(...params: any[]) => ReturnType<F>>(fn: F, delay: number): F {
	let timeoutID: ReturnType<typeof setTimeout>
	return function(...args: unknown[]) {
		clearTimeout(timeoutID)
		timeoutID = setTimeout(() => fn(args), delay)
	} as F
}

export function value<T>(value: T | (() => T)): T {
	if (typeof value === 'function') {
		return (value as any)?.() as T
	}

	return value
}

export function when<T, D>(condition: any, data: T, _default?: D): T | D | undefined {
	if (!condition) {
		return _default
	}

	return data
}

interface MergeOptions {
	overwriteArray?: boolean
	arrayMerge?: (target: any[], source: any[]) => any[]
}

export function merge<T>(x: Partial<T>, y: Partial<T>, options: MergeOptions = {}): T {
	const arrayMerge = typeof options?.arrayMerge === 'function'
		? options.arrayMerge
		: options?.overwriteArray !== false
			? (_: any, s: any) => s
			: undefined

	return baseMerge(x, y, { arrayMerge })
}

export function removeTrailingSlash(string: string): string {
	return string.replace(/\/+$/, '')
}
