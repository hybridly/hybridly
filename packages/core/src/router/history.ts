import { debounce, debug } from '@hybridly/utils'
import { SCROLL_REGION_ATTRIBUTE } from '../constants'
import type { InternalRouterContext, RouterContextOptions, Serializer } from '../context'
import { getInternalRouterContext, getRouterContext, setContext } from '../context'
import { runHooks } from '../plugins'
import { saveScrollPositions } from '../scroll'
import { makeUrl } from '../url'
import { navigate } from './router'

type SerializedContext = Omit<InternalRouterContext, 'adapter' | 'serializer' | 'plugins' | 'hooks' | 'axios' | 'routes'>

/** Puts the given context into the history state. */
export function setHistoryState(options: HistoryOptions = {}) {
	if (!window?.history) {
		throw new Error('The history API is not available, so Hybridly cannot operate.')
	}

	const context = getRouterContext()
	const method = options.replace
		? 'replaceState'
		: 'pushState'

	const serialized = serializeContext(context)

	debug.history('Setting history state:', {
		method,
		context,
		serialized,
	})

	try {
		window.history[method](serialized, '', context.url)
	} catch (error) {
		console.error('Hybridly could not save its current state in the history. This is most likely due to a property being non-serializable, such as a proxy or a reference.')
		throw error
	}
}

/** Gets the current history state if it exists. */
export function getHistoryState<T = any>(key?: string): T {
	const state = key
		? window.history.state?.state?.[key]
		: window.history.state?.state

	return getRouterContext().serializer.unserialize<T>(state)
}

/** Register history-related event listeneners. */
export async function registerEventListeners() {
	const context = getRouterContext()
	debug.history('Registering [popstate] and [scroll] event listeners.')

	// Popstate is for catching back and forward navigations. We want
	// to imitate native browser behavior while keeping the SPA feeling.
	window?.addEventListener('popstate', async(event) => {
		debug.history('Navigation detected (popstate event). State:', { state: event.state })

		await runHooks('backForward', {}, event.state, context)

		// If there is no state in this history entry, we come from the user
		// replacing the URL manually. In this case, we want to restore everything
		// like it was. We need to copy the hash if any and restore the scroll positions.
		if (!event.state) {
			debug.history('There is no state. Adding hash if any and restoring scroll positions.')

			return await navigate({
				payload: {
					...context,
					url: makeUrl(context.url, { hash: window.location.hash }).toString(),
				},
				preserveScroll: true,
				preserveState: true,
				replace: true,
			})
		}

		// If the history entry has been tempered with, we want
		// to use it. We swap the components accordingly.
		await navigate({
			payload: event.state,
			preserveScroll: true,
			preserveState: !!getInternalRouterContext().dialog || !!event.state.dialog,
			updateHistoryState: false,
			isBackForward: true,
		})
	})

	// On scroll, we want to save the positions of all scrollbars.
	// This is needed in order to restore them upon navigation.
	window?.addEventListener('scroll', (event) => debounce(() => {
		if ((event?.target as Element)?.hasAttribute?.(SCROLL_REGION_ATTRIBUTE)) {
			saveScrollPositions()
		}
	}, 100), true)
}

/** Checks if the current navigation was made by going back or forward. */
export function isBackForwardNavigation(): boolean {
	if (!window.history.state) {
		return false
	}

	return (window.performance?.getEntriesByType('navigation').at(0) as PerformanceNavigationTiming)?.type === 'back_forward'
}

/** Handles a navigation which was going back or forward. */
export async function handleBackForwardNavigation(): Promise<void> {
	debug.router('Handling a back/forward navigation.')
	window.history.state.version = getRouterContext().version

	await navigate({
		payload: window.history.state,
		preserveScroll: true,
		preserveState: false,
		updateHistoryState: false,
		isBackForward: true,
	})
}

/** Saves a value into the current history state. */
export function remember<T = any>(key: string, value: T): void {
	debug.history(`Remembering key "${key}" with value`, value)

	// We avoid propagation in order to not trigger a recursive
	// render by telling the adapter we updated.
	setContext({
		state: {
			...getRouterContext().state,
			[key]: value,
		},
	}, { propagate: false })

	setHistoryState({ replace: true })
}

/** Gets a value saved into the current history state. */
export function getKeyFromHistory<T = any>(key: string): T {
	return getHistoryState<T>(key)
}

/** Serializes the context so it can be written to the history state. */
export function serializeContext(context: InternalRouterContext): SerializedContext {
	return {
		url: context.url,
		version: context.version,
		view: context.serializer.serialize(context.view),
		dialog: context.dialog,
		scrollRegions: context.scrollRegions,
		state: context.serializer.serialize(context.state),
	}
}

export function createSerializer(options: RouterContextOptions): Serializer {
	if (options.serializer) {
		return options.serializer
	}

	return {
		serialize: (view) => JSON.parse(JSON.stringify(view)),
		unserialize: (state) => state,
	}
}

export interface HistoryOptions {
	/**
	 * Replace the current history entry instead of pushing a new entry.
	 * @default false
	 */
	replace?: boolean
}
