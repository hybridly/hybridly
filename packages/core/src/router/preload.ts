import { debug } from '@hybridly/utils'
import type { AxiosResponse } from 'axios'
import { getInternalRouterContext, getRouterContext } from '../context'
import { makeUrl } from '../url'
import type { HybridRequestOptions } from './types'
import { performHybridRequest, isHybridResponse } from './router'

/**
 * Checks if there is a preloaded request for the given URL.
 */
export function isPreloaded(targetUrl: URL): boolean {
	const context = getInternalRouterContext()
	return context.preloadCache.has(targetUrl.toString()) ?? false
}

/**
 * Gets and erases a preloaded request.
 */
export function pullPreloadedRequest(targetUrl: URL): AxiosResponse | undefined {
	const response = getPreloadedRequest(targetUrl)
	discardPreloadedRequest(targetUrl)

	return response
}

/**
 * Gets the response of a preloaded request.
 */
export function getPreloadedRequest(targetUrl: URL): AxiosResponse | undefined {
	const context = getInternalRouterContext()
	return context.preloadCache.get(targetUrl.toString())
}

/**
 * Stores the response of a preloaded request.
 */
export function storePreloadRequest(targetUrl: URL, response: AxiosResponse) {
	const context = getInternalRouterContext()
	context.preloadCache.set(targetUrl.toString(), response)
}

/**
 * Discards a preloaded request.
 */
export function discardPreloadedRequest(targetUrl: URL) {
	const context = getInternalRouterContext()
	return context.preloadCache.delete(targetUrl.toString())
}

/** Preloads a hybrid request. */
export async function performPreloadRequest(options: HybridRequestOptions): Promise<boolean> {
	const context = getRouterContext()
	const url = makeUrl(options.url ?? context.url)

	if (isPreloaded(url)) {
		debug.router('This request is already preloaded.')
		return false
	}

	if (context.pendingNavigation) {
		debug.router('A navigation is pending, preload aborted.')
		return false
	}

	if (options.method !== 'GET') {
		debug.router('Cannot preload non-GET requests.')
		return false
	}

	debug.router(`Preloading response for [${url.toString()}]`)

	try {
		const response = await performHybridRequest(url, options)

		if (!isHybridResponse(response)) {
			debug.router('Preload result was invalid.')
			return false
		}

		storePreloadRequest(url, response)
		return true
	} catch (error) {
		debug.router('Preloading failed.')
		return false
	}
}
