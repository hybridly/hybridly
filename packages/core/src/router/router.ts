import type { AxiosProgressEvent, AxiosResponse } from 'axios'
import { showResponseErrorModal, match, merge, when, debug, random, hasFiles, objectToFormData } from '@hybridly/utils'
import { ERROR_BAG_HEADER, EXCEPT_DATA_HEADER, EXTERNAL_NAVIGATION_HEADER, ONLY_DATA_HEADER, PARTIAL_COMPONENT_HEADER, HYBRIDLY_HEADER, VERSION_HEADER, DIALOG_KEY_HEADER, DIALOG_REDIRECT_HEADER } from '../constants'
import { NotAHybridResponseError, NavigationCancelledError } from '../errors'
import type { InternalRouterContext, RouterContextOptions } from '../context'
import { getRouterContext, initializeContext, payloadFromContext, setContext } from '../context'
import { handleExternalNavigation, isExternalResponse, isExternalNavigation, performExternalNavigation, navigateToExternalUrl } from '../external'
import { resetScrollPositions, restoreScrollPositions, saveScrollPositions } from '../scroll'
import type { UrlResolvable } from '../url'
import { sameHashes, fillHash, makeUrl, normalizeUrl, sameUrls } from '../url'
import { runHooks } from '../plugins'
import { generateRouteFromName, getRouteDefinition } from '../routing/route'
import { closeDialog } from '../dialog'
import { setHistoryState, isBackForwardNavigation, handleBackForwardNavigation, registerEventListeners, getHistoryState, getKeyFromHistory, remember } from './history'
import type { ConditionalNavigationOption, Errors, ComponentNavigationOptions, NavigationOptions, Router, HybridRequestOptions, HybridPayload, NavigationResponse, Method } from './types'

/**
 * The hybridly router.
 * This is the core function that you can use to navigate in
 * your application. Make sure the routes you call return a
 * hybrid response, otherwise you need to call `external`.
 *
 * @example
 * router.get('/posts/edit', { post })
 */
export const router: Router = {
	abort: async() => getRouterContext().pendingNavigation?.controller.abort(),
	active: () => !!getRouterContext().pendingNavigation,
	navigate: async(options) => await performHybridNavigation(options),
	reload: async(options) => await performHybridNavigation({ preserveScroll: true, preserveState: true, ...options }),
	get: async(url, options = {}) => await performHybridNavigation({ ...options, url, method: 'GET' }),
	post: async(url, options = {}) => await performHybridNavigation({ preserveState: true, ...options, url, method: 'POST' }),
	put: async(url, options = {}) => await performHybridNavigation({ preserveState: true, ...options, url, method: 'PUT' }),
	patch: async(url, options = {}) => await performHybridNavigation({ preserveState: true, ...options, url, method: 'PATCH' }),
	delete: async(url, options = {}) => await performHybridNavigation({ preserveState: true, ...options, url, method: 'DELETE' }),
	local: async(url, options = {}) => await performLocalNavigation(url, options),
	external: (url, data = {}) => navigateToExternalUrl(url, data),
	to: async(name, parameters, options) => {
		const url = generateRouteFromName(name, parameters)
		const method = getRouteDefinition(name).method.at(0)
		return await performHybridNavigation({ url, ...options, method })
	},
	dialog: {
		close: (options) => closeDialog(options),
	},
	history: {
		get: (key) => getKeyFromHistory(key),
		remember: (key, value) => remember(key, value),
	},
}

/** Creates the hybridly router. */
export async function createRouter(options: RouterContextOptions): Promise<InternalRouterContext> {
	await initializeContext(options)

	return await initializeRouter()
}

/** Performs every action necessary to make a hybrid navigation. */
export async function performHybridNavigation(options: HybridRequestOptions): Promise<NavigationResponse> {
	const navigationId = random()
	const context = getRouterContext()
	debug.router('Making a hybrid navigation:', { context, options, navigationId })

	try {
		// Sets the method if not specifically defined.
		if (!options.method) {
			debug.router('Setting method to GET because none was provided.')
			options.method = 'GET'
		}

		// Force uppercase method because we accept lowercase methods,
		// *angry look at Hassan*
		options.method = options.method.toUpperCase() as Method

		// If applicable, converts the data to a `FormData` object.
		if ((hasFiles(options.data) || options.useFormData) && !(options.data instanceof FormData)) {
			options.data = objectToFormData(options.data)
			debug.router('Converted data to FormData.', options.data)

			// Automatically spoofs the method of the form, because it is not possible to
			// send a PUT, PATCH or DELETE request with files in them.
			if (options.method && ['PUT', 'PATCH', 'DELETE'].includes(options.method) && options.spoof !== false) {
				debug.router(`Automatically spoofing method ${options.method}.`)
				options.data.append('_method', options.method)
				options.method = 'POST'
			}
		}

		// Converts data to query parameters if the method is GET
		// and some non-FormData data is provided.
		if (!(options.data instanceof FormData) && options.method === 'GET' && Object.keys(options.data ?? {}).length) {
			debug.router('Transforming data to query parameters.', options.data)
			options.url = makeUrl(options.url ?? context.url, {
				query: options.data,
			})
			options.data = {}
		}

		// Before anything else, we fire the "before" event to make sure
		// there was no user-specified handler returning "false".
		if (!await runHooks('before', options.hooks, options, context)) {
			debug.router('"before" event returned false, aborting the navigation.')
			throw new NavigationCancelledError('The navigation was cancelled by the "before" event.')
		}

		// Abort any active navigation.
		if (context.pendingNavigation) {
			debug.router('Aborting current navigation.', context.pendingNavigation)
			context.pendingNavigation?.controller?.abort()
		}

		// Before making the navigation, we need to make sure the scroll positions are
		// saved, so we can restore them later.
		saveScrollPositions()

		// If the URL has transformation options, apply them before using the URL.
		if (options.url && options.transformUrl) {
			options.url = makeUrl(options.url, options.transformUrl)
		}

		const targetUrl = makeUrl(options.url ?? context.url)
		const abortController = new AbortController()

		// A navigation is being made, we need to add it to the context so it
		// can be reused later on.
		setContext({
			pendingNavigation: {
				id: navigationId,
				url: targetUrl,
				controller: abortController,
				status: 'pending',
				options,
			},
		})

		await runHooks('start', options.hooks, context)
		debug.router('Making request with axios.')

		const response = await context.axios.request({
			url: targetUrl.toString(),
			method: options.method,
			data: options.method === 'GET' ? {} : options.data,
			params: options.method === 'GET' ? options.data : {},
			signal: abortController.signal,
			headers: {
				...options.headers,
				...(context.dialog ? { [DIALOG_KEY_HEADER]: context.dialog!.key } : {}),
				...(context.dialog ? { [DIALOG_REDIRECT_HEADER]: context.dialog!.redirectUrl ?? '' } : {}),
				...when(options.only !== undefined || options.except !== undefined, {
					[PARTIAL_COMPONENT_HEADER]: context.view.component,
					...when(options.only, { [ONLY_DATA_HEADER]: JSON.stringify(options.only) }, {}),
					...when(options.except, { [EXCEPT_DATA_HEADER]: JSON.stringify(options.except) }, {}),
				}, {}),
				...when(options.errorBag, { [ERROR_BAG_HEADER]: options.errorBag }, {}),
				...when(context.version, { [VERSION_HEADER]: context.version }, {}),
				[HYBRIDLY_HEADER]: true,
				'X-Requested-With': 'XMLHttpRequest',
				'Accept': 'text/html, application/xhtml+xml',
			},
			validateStatus: () => true,
			onUploadProgress: async(event: AxiosProgressEvent) => {
				await runHooks('progress', options.hooks, {
					event,
					percentage: Math.round(event.loaded / (event.total ?? 0) * 100),
				}, context)
			},
		})

		await runHooks('data', options.hooks, response, context)

		// An external response is a hybrid response that wants a full page
		// load to a requested URL. It may be the same URL, in which
		// case a full page refresh will be performed.
		if (isExternalResponse(response)) {
			debug.router('The response is explicitely external.')
			await performExternalNavigation({
				url: fillHash(targetUrl, response.headers[EXTERNAL_NAVIGATION_HEADER]!),
				preserveScroll: options.preserveScroll === true,
			})

			return { response }
		}

		// An invalid response is a response that do not declare itself via
		// the protocole header.
		// In such cases, we want to throw to handler it later.
		if (!isHybridResponse(response)) {
			throw new NotAHybridResponseError(response)
		}

		// At this point, we know the response respects the hybridly protocol.
		debug.router('The response respects the Hybridly protocol.')
		const payload = response.data as HybridPayload

		// If the navigation was asking for specific properties, we ensure that the
		// new request object contains the properties of the current view context,
		// because the back-end sent back only the required properties.
		if ((options.only?.length ?? options.except?.length) && payload.view.component === context.view.component) {
			debug.router(`Merging ${options.only ? '"only"' : '"except"'} properties.`, payload.view.properties)
			payload.view.properties = merge(context.view.properties, payload.view.properties)
			debug.router('Merged properties:', payload.view.properties)
		}

		// If everything was according to the plan, we can make our navigation and
		// update the context. Underlying adapters get the updated data.
		await navigate({
			payload: {
				...payload,
				url: fillHash(targetUrl, payload.url),
			},
			preserveScroll: options.preserveScroll,
			preserveState: options.preserveState,
			preserveUrl: options.preserveUrl,
			replace: options.replace === true || sameHashes(payload.url, window.location.href) || options.preserveUrl,
		})

		// If the new view's properties has errors, userland expects an event
		// with said errors to be emitted. However, errors can be scoped with
		// an error bag, and if the given error bag is missing, the event data
		// will be empty.
		if (Object.keys(context.view.properties.errors ?? {}).length > 0) {
			const errors = (() => {
				if (options.errorBag && typeof context.view.properties.errors === 'object') {
					return (context.view.properties.errors as any)[options.errorBag] ?? {}
				}

				return context.view.properties.errors
			})() as Errors

			debug.router('The request returned validation errors.', errors)
			setContext({
				pendingNavigation: {
					...context.pendingNavigation!,
					status: 'error',
				},
			})

			await runHooks('error', options.hooks, errors, context)
		} else {
			setContext({
				pendingNavigation: {
					...context.pendingNavigation!,
					status: 'success',
				},
			})

			await runHooks('success', options.hooks, payload, context)
		}

		return { response }
	//
	} catch (error: any) {
		await match(error.constructor.name, {
			NavigationCancelledError: async() => {
				debug.router('The request was cancelled through the "before" hook.', error)
				await runHooks('abort', options.hooks, context)
			},
			AbortError: async() => {
				debug.router('The request was aborted.', error)
				await runHooks('abort', options.hooks, context)
			},
			NotAHybridResponseError: async() => {
				debug.router('The request was not hybridly.')
				console.error(error)
				await runHooks('invalid', options.hooks, error, context)
				if (context.responseErrorModals) {
					showResponseErrorModal(error.response.data)
				}
			},
			default: async() => {
				if (error?.name === 'CanceledError') {
					debug.router('The request was cancelled.', error)
					await runHooks('abort', options.hooks, context)
				} else {
					debug.router('An unknown error occured.', error)
					console.error(error)
					await runHooks('exception', options.hooks, error, context)
				}
			},
		})

		await runHooks('fail', options.hooks, context)

		return {
			error: {
				type: error.constructor.name,
				actual: error,
			},
		}
	} finally {
		debug.router('Ending navigation.')
		await runHooks('after', options.hooks, context)

		// If the navigation is pending, it means another one has started,
		// and if we clean it up, it could lead to issues after the request is done.
		if (context.pendingNavigation?.id === navigationId) {
			setContext({ pendingNavigation: undefined })
		}
	}
}

/** Checks if the response contains a hybrid header. */
export function isHybridResponse(response: AxiosResponse): boolean {
	return !!response?.headers[HYBRIDLY_HEADER]
}

/**
 * Makes an internal navigation that swaps the view and updates the context.
 * @internal This function is meant to be used internally.
 */
export async function navigate(options: NavigationOptions) {
	const context = getRouterContext()
	debug.router('Making an internal navigation:', { context, options })

	await runHooks('navigating', {}, options, context)

	// If no request was given, we use the current context instead.
	options.payload ??= payloadFromContext()

	function evaluateConditionalOption<T extends boolean | string>(option?: ConditionalNavigationOption<T>) {
		return typeof option === 'function'
			? option(options)
			: option
	}

	const shouldPreserveState = evaluateConditionalOption(options.preserveState)
	const shouldPreserveScroll = evaluateConditionalOption(options.preserveScroll)
	const shouldReplaceHistory = evaluateConditionalOption(options.replace)
	const shouldReplaceUrl = evaluateConditionalOption(options.preserveUrl)

	// If the navigation was asking to preserve the current state, we also need to
	// update the context's state from the history state.
	if (shouldPreserveState && getHistoryState() && options.payload.view.component === context.view.component) {
		debug.history('Setting the history from this entry into the context.')
		setContext({ state: getHistoryState() })
	}

	// If the navigation required the URL to be preserved, we skip its update
	// by replacing the payload URL with the current context URL.
	if (shouldReplaceUrl) {
		debug.router(`Preserving the current URL (${context.url}) instead of navigating to ${options.payload.url}`)
		options.payload!.url = context.url
	}

	// We merge the new request into the current context. That will replace
	// view, dialog, url and version, so the context is in sync with the
	// navigation that took place.
	setContext({
		...options.payload,
		state: {},
	})

	// History state must be updated to preserve the expected, native browser behavior.
	// However, in some cases, we just want to swap the views without making an
	// actual navigation. Additionally, we don't want to actually push a new state
	// when navigating to the same URL.
	if (options.updateHistoryState !== false) {
		debug.router(`Target URL is ${context.url}, current window URL is ${window.location.href}.`, { shouldReplaceHistory })
		setHistoryState({ replace: shouldReplaceHistory })
	}

	// Then, we swap the view.
	const viewComponent = await context.adapter.resolveComponent(context.view.component)
	debug.router(`Component [${context.view.component}] resolved to:`, viewComponent)
	await context.adapter.onViewSwap({
		component: viewComponent,
		dialog: context.dialog,
		properties: options.payload?.view.properties,
		preserveState: shouldPreserveState,
	})

	if (options.isBackForward) {
		restoreScrollPositions()
	} else if (!shouldPreserveScroll) {
		resetScrollPositions()
	}

	await runHooks('navigated', {}, options, context)
}

/** Initializes the router by reading the context and registering events if necessary. */
async function initializeRouter(): Promise<InternalRouterContext> {
	const context = getRouterContext()

	if (isBackForwardNavigation()) {
		handleBackForwardNavigation()
	} else if (isExternalNavigation()) {
		handleExternalNavigation()
	} else {
		debug.router('Handling the initial page navigation.')

		// If we navigated to somewhere with a hash, we need to update the context
		// to add said hash because it was initialized without it.
		setContext({
			url: makeUrl(context.url, { hash: window.location.hash }).toString(),
		})

		await navigate({
			preserveState: true,
			replace: sameUrls(context.url, window.location.href),
		})
	}

	registerEventListeners()

	await runHooks('ready', {}, context)

	return context
}

/** Performs a local navigation to the given component without a round-trip. */
export async function performLocalNavigation(targetUrl: UrlResolvable, options?: ComponentNavigationOptions) {
	const context = getRouterContext()
	const url = normalizeUrl(targetUrl)

	return await navigate({
		...options,
		payload: {
			version: context.version,
			dialog: options?.dialog === false ? undefined : (options?.dialog ?? context.dialog),
			url,
			view: {
				component: options?.component ?? context.view.component,
				properties: options?.properties ?? {},
			},
		},
	})
}
