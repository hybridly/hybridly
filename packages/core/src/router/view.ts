import { debug, merge } from '@hybridly/utils'
import { resetScrollPositions, restoreScrollPositions } from '../scroll'
import { runHooks } from '../plugins'
import { getRouterContext, payloadFromContext, setContext } from '../context'
import type { UrlResolvable } from '../url'
import { normalizeUrl } from '../url'
import { getHistoryMemo, setHistoryState } from './history'
import type { ComponentNavigationOptions, ConditionalNavigationOption, HybridPayload, InternalNavigationOptions } from './types'
import { performHybridNavigation } from './request/request'

/**
 * Makes an internal navigation that swaps the view and updates the context.
 * @internal
 */
export async function navigate(options: InternalNavigationOptions) {
	const context = getRouterContext()

	// Since there is no other way to know prior to the navigation actually being made,
	// we mutate `options` here to add whether there is a dialog or not.
	options.hasDialog ??= !!options.payload?.dialog

	debug.router('Making an internal navigation:', { context, options })
	await runHooks('navigating', {}, options, context)

	// If no request was given, we use the current context instead.
	options.payload ??= payloadFromContext()
	options.payload.view ??= payloadFromContext().view
	options.payload.view.properties = options.properties ?? options.payload.view.properties

	function evaluateConditionalOption<T extends boolean | string>(option?: ConditionalNavigationOption<T>) {
		return typeof option === 'function'
			? option(options)
			: option
	}

	const shouldPreserveState = evaluateConditionalOption(options.preserveState)
	const shouldPreserveScroll = evaluateConditionalOption(options.preserveScroll)
	const shouldReplaceHistory = evaluateConditionalOption(options.replace)
	const shouldReplaceUrl = evaluateConditionalOption(options.preserveUrl)
	const shouldPreserveView = !options.payload.view.component

	// If the navigation was asking to preserve the current state, we also need to
	// update the context's state from the history state.
	if (shouldPreserveState && getHistoryMemo() && options.payload.view.component === context.view.component) {
		debug.history('Setting the memo from this history entry into the current context.')
		setContext({ memo: getHistoryMemo() })
	}

	// If the navigation required the URL to be preserved, we skip its update
	// by replacing the payload URL with the current context URL.
	if (shouldReplaceUrl) {
		debug.router(`Preserving the current URL (${context.url}) instead of navigating to ${options.payload.url}`)
		options.payload!.url = context.url
	}

	// If we didn't receive a component name,
	// we don't swap views and we preserve the url.
	const payload = shouldPreserveView
		? {
			view: {
				component: context.view.component,
				properties: merge(context.view.properties, options.payload.view.properties),
				deferred: context.view.deferred,
				mergeable: context.view.mergeable,
			},
			url: context.url,
			version: options.payload.version,
			dialog: context.dialog,
		} satisfies HybridPayload
		: options.payload

	// We merge the new request into the current context. That will replace
	// view, dialog, url and version, so the context is in sync with the
	// navigation that took place.
	setContext({ ...payload, memo: {} })

	// History state must be updated to preserve the expected, native browser behavior.
	// However, in some cases, we just want to swap the views without making an
	// actual navigation. Additionally, we don't want to actually push a new state
	// when navigating to the same URL.
	if (options.updateHistoryState !== false) {
		debug.router(`Target URL is ${context.url}, current window URL is ${window.location.href}.`, { shouldReplaceHistory })
		setHistoryState({ replace: shouldReplaceHistory })
	}

	// If there are deferred properties, we handle them
	// by making a partial-reload after the view component has mounted
	if (Object.entries(context.view.deferred ?? {}).length) {
		debug.router('Request has deferred properties, queueing a partial reload:', context.view.deferred)
		context.adapter.executeOnMounted(async () => {
			return Promise.all(Object.entries(context.view.deferred).map(async ([_, properties]) => {
				await performHybridNavigation({
					preserveScroll: true,
					preserveState: true,
					replace: true,
					async: true,
					only: properties,
				})
			}))
		})
	}

	// Then, we swap the view.
	const viewComponent = !shouldPreserveView
		? await context.adapter.resolveComponent(context.view.component!)
		: undefined

	if (viewComponent) {
		debug.router(`Component [${context.view.component}] resolved to:`, viewComponent)
	}

	await context.adapter.onViewSwap({
		component: viewComponent,
		dialog: context.dialog,
		properties: options.payload?.view?.properties,
		preserveState: shouldPreserveState,
		onMounted: (hookOptions) => runHooks('mounted', {}, { ...options, ...hookOptions }, context),
	})

	if (options.type === 'back-forward') {
		restoreScrollPositions()
	} else if (!shouldPreserveScroll) {
		resetScrollPositions()
	}

	await runHooks('navigated', {}, options, context)
}

/** Performs a local navigation to the given component without a round-trip. */
export async function performLocalNavigation(targetUrl: UrlResolvable, options?: ComponentNavigationOptions) {
	const context = getRouterContext()
	const url = normalizeUrl(targetUrl)

	return await navigate({
		...options,
		type: 'local',
		payload: {
			version: context.version,
			dialog: options?.dialog === false ? undefined : (options?.dialog ?? context.dialog),
			url,
			view: {
				component: options?.component ?? context.view.component,
				properties: options?.properties ?? {},
				deferred: {},
				mergeable: [],
			},
		},
	})
}
