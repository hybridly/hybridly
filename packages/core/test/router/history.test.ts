import { afterAll, beforeAll, beforeEach, describe, expect, it, vi } from 'vitest'
import { createSerializer, setHistoryState } from '../../src/router/history'
import { fakeRouterContext, makeRouterContextOptions, returnsArgs } from '../utils'

beforeEach(async() => {
	await fakeRouterContext()
})

describe('setHistoryState', () => {
	beforeAll(() => {
		window.location.href = 'https://localhost'
	})

	afterAll(() => {
		vi.clearAllMocks()
	})

	it('pushes the context into the history', () => {
		const spy = vi.spyOn(window.history, 'pushState').mockImplementation(returnsArgs)

		setHistoryState()

		expect(spy).toHaveBeenCalledOnce()
		expect(spy.mock.results).toMatchSnapshot('context')
	})

	it('replaces the context in the current history state', () => {
		const spy = vi.spyOn(window.history, 'replaceState').mockImplementation(returnsArgs)

		setHistoryState({ replace: true })

		expect(spy).toHaveBeenCalledOnce()
		expect(spy.mock.results).toMatchSnapshot('context')
	})
})

describe('serializer', () => {
	it('serializes and unserializes complex objects', () => {
		const options = makeRouterContextOptions()
		const serializer = createSerializer(options)
		const data = {
			set: new Set([1, 2, 3, 'string', { foo: 'bar' }]),
			foo: 'bar',
			nested: {
				foo: 'bar',
				set: new Set(['a', 'b', 1, 2, 3, 'c']),
			},
			nestedTwice: {
				foo: {
					set: new Set([1, 2, 3, 'soleil']),
				},
			},
			empty: undefined,
			map: new Map([
				['a', 'test1'],
				['b', 'test2'],
			]),
		}

		const serialized = serializer.serialize(data)

		expect(serialized).toMatchSnapshot()
		expect(data).toStrictEqual(serializer.unserialize(serialized))
	})
})
