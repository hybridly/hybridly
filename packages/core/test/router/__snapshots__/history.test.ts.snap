// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`serializer > serializes and unserializes complex objects 1`] = `"{\\"json\\":{\\"set\\":[1,2,3,\\"string\\",{\\"foo\\":\\"bar\\"}],\\"foo\\":\\"bar\\",\\"nested\\":{\\"foo\\":\\"bar\\",\\"set\\":[\\"a\\",\\"b\\",1,2,3,\\"c\\"]},\\"nestedTwice\\":{\\"foo\\":{\\"set\\":[1,2,3,\\"soleil\\"]}},\\"empty\\":null,\\"map\\":[[\\"a\\",\\"test1\\"],[\\"b\\",\\"test2\\"]]},\\"meta\\":{\\"values\\":{\\"set\\":[\\"set\\"],\\"nested.set\\":[\\"set\\"],\\"nestedTwice.foo.set\\":[\\"set\\"],\\"empty\\":[\\"undefined\\"],\\"map\\":[\\"map\\"]}}}"`;

exports[`setHistoryState > pushes the context into the history > context 1`] = `
[
  {
    "type": "return",
    "value": [
      "{\\"json\\":{\\"url\\":\\"https://localhost/\\",\\"version\\":\\"abc123\\",\\"view\\":{\\"component\\":\\"default.view\\",\\"properties\\":{}},\\"dialog\\":null,\\"scrollRegions\\":[],\\"memo\\":{}},\\"meta\\":{\\"values\\":{\\"dialog\\":[\\"undefined\\"]}}}",
      "",
      "https://localhost/",
    ],
  },
]
`;

exports[`setHistoryState > replaces the context in the current history state > context 1`] = `
[
  {
    "type": "return",
    "value": [
      "{\\"json\\":{\\"url\\":\\"https://localhost/\\",\\"version\\":\\"abc123\\",\\"view\\":{\\"component\\":\\"default.view\\",\\"properties\\":{}},\\"dialog\\":null,\\"scrollRegions\\":[],\\"memo\\":{}},\\"meta\\":{\\"values\\":{\\"dialog\\":[\\"undefined\\"]}}}",
      "",
      "https://localhost/",
    ],
  },
]
`;
